ikarus/finiteelements/mechanics/materials/hyperelastic/interface.hh

```diff
@@ -0,0 +1,330 @@
+  auto rebind() const {
+    auto reboundDEV = dev_.template rebind<STO>();
+    return Hyperelastic<decltype(reboundDEV), VOL>(reboundDEV, vol_);
+  }
+
+private:
+  DEV dev_;
+  VOL vol_;
+
+  inline static constexpr auto dimensionRange() { return Dune::range(dim); }
+
+  template <typename ST>
+  Eigen::Matrix<ST, dim, dim> transformDeviatoricStresses(const Eigen::Vector<ST, dim>& principalStress,
+                                                          const Eigen::Matrix<ST, dim, dim>& N) const {
+    return (N * principalStress.asDiagonal() * N.transpose()).eval();
+  }
+
+  template <typename ST>
+  Eigen::Matrix<ST, dim, dim> transformVolumetricStresses(ST Uprime, const auto& C, ST J) const {
+    return J * Uprime * C.inverse();
+  }
+
+  template <typename ST>
+  auto transformDeviatoricTangentModuli(const Eigen::TensorFixedSize<ST, Eigen::Sizes<dim, dim, dim, dim>>& L,
+                                        const Eigen::Matrix<ST, dim, dim>& N) const {
+    Eigen::TensorFixedSize<ST, Eigen::Sizes<dim, dim, dim, dim>> moduli{};
+    moduli.setZero();
+    auto indexList = std::array<Eigen::Index, 2>({dim, dim});
+
+    for (const auto i : dimensionRange())
+      for (const auto k : dimensionRange()) {
+        // First term: L[i, i, k, k] * ((N[i] ⊗ N[i]) ⊗ (N[k] ⊗ N[k]))
+        auto NiNi = tensorView(dyadic(N.col(i).eval(), N.col(i).eval()), indexList);
+        auto NkNk = tensorView(dyadic(N.col(k).eval(), N.col(k).eval()), indexList);
+
+        moduli += L(i, i, k, k) * dyadic(NiNi, NkNk);
+
+        // Second term (only if i != k): L[i, k, i, k] * (N[i] ⊗ N[k] ⊗ (N[i] ⊗ N[k] + N[k] ⊗ N[i]))
+        if (i != k) {
+          auto NiNk = tensorView(dyadic(N.col(i).eval(), N.col(k).eval()), indexList);
+          auto NkNi = tensorView(dyadic(N.col(k).eval(), N.col(i).eval()), indexList);
+
+          moduli += L(i, k, i, k) * dyadic(NiNk, NiNk + NkNi);
+        }
+      }
+
+    return moduli;
+  }
+
+  template <typename ST>
+  auto transformVolumetricTangentModuli(const ST& Uprime, const ST& Uprimeprime, const auto& C, ST J) const {
+    const auto invC  = C.inverse().eval();
+    const auto CTinv = tensorView(invC, std::array<Eigen::Index, 2>({3, 3}));
+
+    Eigen::TensorFixedSize<ST, Eigen::Sizes<3, 3, 3, 3>> moduli =
+        (J * ((Uprime + J * Uprimeprime) * dyadic(CTinv, CTinv) -
+              (2 * Uprime * symTwoSlots(fourthOrderIKJL(invC, invC), {2, 3}))))
+            .eval();
+
+    return moduli;
+  }
+
+  template <typename Derived>
+  auto principalStretches(const Eigen::MatrixBase<Derived>& Craw, int options = Eigen::ComputeEigenvectors) const {
+    auto C = Impl::maybeFromVoigt(Craw);
+    return Impl::principalStretches<typename Derived::Scalar>(C, options);
+  }
+
+  template <typename Derived, typename ST>
+  ST detF(const Eigen::MatrixBase<Derived>& C, const Eigen::Vector<ST, 3>& lambda) const {
+    if constexpr (isAutoDiff) {
+      const auto detC = sqrt(C.derived().eval().determinant());
```

Comment:
I investigated the reason for this:

In https://eigen.tuxfamily.org/dox/Tridiagonalization_8h_source.html in line 477, there is an early exit if the 2,0 entry of the matrix is zero. then an early exit is performed, which is fine for doubles, but for the autodiff type there information gets lost. But if we use compute direct there the sqrt is taken of an entry that can be zero which would als o create nans for us. Therefore, the implementration is fine but can you write my statement as a comment to have this stored for later, if somebody wonders why this is so strange.

Reply:
Thanks for the investigation. Added comment